flax/nnx/transforms/compilation.py

```diff
@@ -405,14 +471,55 @@ def __get__(self, obj, objtype=None):
     return functools.partial(self, obj)
 
   def _get_pure_args_kwargs(self, args, kwargs):
+    # Build a prefix for positional arguments that inserts `None` for static
+    # positions so that prefix broadcasting matches the full argument tuple.
+    def _build_positional_prefix():
+      if self.in_shardings is None:
+        return None
+      # Determine number of positional args and number of dynamic args.
+      num_args = len(args)
+      static_set = set(self._static_positions)
+      num_dynamic = num_args - sum(1 for i in range(num_args) if i in static_set)
+
+      # Expand user-provided in_shardings to a per-dynamic-arg sequence.
+      if isinstance(self.in_shardings, (tuple, list)):
+        dyn_specs = list(self.in_shardings)
+        # If user provided a shorter sequence (e.g., as a prefix), do not try
+        # to out-broadcast here; rely on extract/jax to handle deeper trees.
+        # When lengths mismatch with expected dynamic count, we will not index
+        # past available entries.
+      else:
+        # Single spec: broadcast to all dynamic args
+        dyn_specs = [self.in_shardings] * max(num_dynamic, 0)
+
+      # Now interleave dyn_specs into a full positional prefix, inserting
+      # None for static positions.
+      full_prefix: list[tp.Any] = []
+      dyn_idx = 0
+      for i in range(num_args):
```

Comment:
It might be slightly cleaner to use an iterator here. 
```python
shard_iter = iter(dyn_specs)
tuple(None if i in static_set else next(dyn_specs) for i in range(num_args))
```